kerrighed-2.4.4/man/krgcapset.1
'\" t
.\"     Title: krgcapset
.\"    Author: [see the "Author" section]
.\" Generator: DocBook XSL Stylesheets v1.75.2 <http://docbook.sf.net/>
.\"      Date: 01/29/2010
.\"    Manual: [FIXME: manual]
.\"    Source: [FIXME: source]
.\"  Language: English
.\"
.TH "KRGCAPSET" "1" "01/29/2010" "[FIXME: source]" "[FIXME: manual]"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
krgcapset \- Set or modify capabilities of a process\&.
.SH "SYNOPSIS"
.HP \w'\fBkrgcapset\fR\ 'u
\fBkrgcapset\fR [\-h | \-\-help]
.HP \w'\fBkrgcapset\fR\ 'u
\fBkrgcapset\fR [\-s | \-\-show]
.HP \w'\fBkrgcapset\fR\ 'u
\fBkrgcapset\fR [\-f | \-\-force] [\-k | \-\-pid] {SET {[+\ |\ \-]\ CAPABILITY\ LIST  | OCTAL\ VALUE}...}
.SH "SET"
.PP
.PP
\-e, \-\-effective
.RS 4
Set up
\fBeffective\fR
capabilities\&.
.RE
.PP
\-p, \-\-permitted
.RS 4
Set up
\fBpermitted\fR
capabilities\&.
.RE
.PP
\-d, \-\-inheritable\-effective
.RS 4
Set up
\fBinheritable effective\fR
capabilities\&.
.RE
.PP
\-i, \-\-inheritable\-permitted
.RS 4
Set up
\fBinheritable permitted\fR
capabilities\&.
.RE
.SH "DESCRIPTION"
.PP
Kerrighed provides a system of
\fIcapabilities\fR, which allows administrator and users to define the way their processes behave in term of SSI mechanisms\&. Kerrighed cluster wide mechanisms are divided into distinct units that can be independently enabled and disabled\&. Each process has four capability sets:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBEffective\fR,
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBPermitted\fR,
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBInheritable Effective\fR,
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBInheritable Permitted\fR\&.
.RE
.PP
See
\fBkerrighed_capabilities\fR(7) for further details\&.
\fBkrgcapset\fR
modify calling process capabilities\&.
.SH "OPTIONS"
.PP
.PP
\-h, \-\-help
.RS 4
Print help and exit\&.
.RE
.PP
\-s, \-\-show
.RS 4
Show the capabilities of the calling process\&.
.RE
.PP
\-f, \-\-force
.RS 4
Do not ask for confirmation when removing CHANGE_KERRIGHED_CAP\&.
.RE
.SH "EXAMPLES"
.PP
.PP
krgcapset \-\-show
.RS 4
Show capabilities of the calling process\&.
.RE
.PP
krgcapset \-\-effective +CAN_MIGRATE
.RS 4
Add the "CAN_MIGRATE" effective capability to the calling process\&.
.RE
.PP
krgcapset \-\-effective \-CAN_MIGRATE
.RS 4
Remove the "CAN_MIGRATE" effective capability to the calling process\&.
.RE
.PP
krgcapset \-\-effective 07
.RS 4
Set effective capabilities to CHANGE_KERRIGHED_CAP, USE_CONTAINERS, CAN_MIGRATE\&.
.RE
.SH "AUTHOR"
.PP
Renaud Lottiaux
renaud\&.lottiaux@kerlabs\&.com
.SH "SEE ALSO"
.PP

\m[blue]\fB\fBkerrighed_capabilities\fR(7)\fR\m[]


kerrighed-2.4.4/ChangeLog
Kerrighed version 2.4.4
   Features:
   * Port driver for Realtek R8169 (present on some Atom platforms)
   * Update tg3 driver from Broadcom website. Add support for BCM5784 for instance
   * Upgrade krgadm
   Fixes:
   * Check for admin right for cluster start/shutdown/restart

Kerrighed version 2.4.2
   Fixes:
   * Fix crashes when doing CTRL-C on a process doing memory injection.
   * Fix some random memory incoherence in process distant fork.
   * Fix OOM in injection mechanism on large memory systems.
   * Avoid migration of pages from processes without USE_REMOTE_MEMORY.
   * Fix some crashes during distant fork or migration on loaded memory system.

Kerrighed version 2.4.1
   Fixes:
   * Major performance improvement
   * Fix crashes during migration or distant fork using Lustre.
   * Fix crashes with distributed system V memory segments.
   * Fix dead-lock in the MM sub-system
   * Fix issues in system V message queue identifiers
   * Fix deadlock on 32bit platforms in the RPC layer
   * Fix forgotten processes in /proc/<pid> directories
   * Fix OOM issue using injection mechanism
   * Fix memory allocation failure when big buffers are passed to read or write
   * Various memory leak fixes.
   * Various fixes in the checkpoint/restart sub-system
   * Various other fixes.

Kerrighed version 2.4.0
   Features:
   * Remote memory injection
   * Configuration scheduler
   * Application check-pointing
   * Remote vfork()
   * New RPC transmission engine that highly increases performance

   Fixes:
   * Fix random seg-fault during process migration on 64bits machines.
   * Fix incompatibilities with IPv6 tunnels and other special net devices
   * Various bugfixes in RPCs
   * Fix a strong memory leak
   * Fix wrong attempts to migrate threads when this is not supported
   * Fix random userspace memory access when using fcntl on files remotely open
   * Fix random userspace memory access when using ioctl on files remotely open
   * Fix BUG when the parent of an exiting multi-threaded process is/was remote
   * Fix many socket related bugs in the context of migrated processes
   * Fix many bugs in krgadm (seg-fault, etc)

   Build system:
   * Remove sub-configures
   * Split tools dir into scripts/ man/ and tools/
   * Create patched kernel source dir in kernel/, no need for --with-kernel option
   * Check architecture compatibility

Kerrighed version 2.3.0
   Fixes:
   * Fix uninitialized anonymous memory mapping
   * Limit network bandwidth consumption due to abusive packet retransmissions
   * Fix root user's ability to kill other's processes when they are remote
   * Fix bug in FAF version of recv and recvfrom
   * Fix user-level deadlocks when interrupting read/write operations with FAF
   * Fix build issues on systems having not sh pointing to bash
   * Fix compile error with debugs and debugfs disabled
   * Fix BUG when processes having multi-threaded children have remote children
   * Fix crash when the scheduler attempts to migrate dead tasks
   * Fix compile error on x86_64
   * Fix unpredictable locking behavior due to the use of a wrong prototype
   * Fix memory leak of slab obj_entry.
   * Fix rare close_faf_file BUG during process migration / distant fork.
   * Fix various signal issues in the FAF mechanism.
   * Fix various signal issues in the communication layer.
   * Fix permission checking when sending a signal to a remote task
   * Fix compilation when building kernel with CONFIG_HIGHMEM_64 on x86_32
   * Add support for disabled CONFIG_COMPAT_VDSO option.
   * Fix system V SHM in SMP mode
   * Fix system V FIFO in SMP mode
   * Fix system V SEM in SMP mode
   * Fix craches related to the use of the recvmsg function

Kerrighed version 2.2.1
   Fixes:
   * Fix crash in fcntl64 with recent libc
   * Fix memory leak in migration and remote fork
   * Fix access to freed memory after migration
   * Fix NUL byte inserted in the first line of /proc/stat
   * Fix memory leaks in FAFed recvmsg() and exit of remote forked processes
   * Fix deadlock when killing non existing (but realistic) pgrp

Kerrighed version 2.2.0
   * SMP support
   * Full distributed IPC support
   * Lots of bug fixes.

Kerrighed version 2.1.0

   * Port to 2.6.20
   * Rewritting of the communication layer on top of TIPC.
   * Lots of bug fixes.

Kerrighed version 2.0.0

   Strong refactoring and janitoring.
   This version is the basis for our future developments.
   Less features than in the previous versions, but more stability and
   much cleaner code.
   Features removed in this version will be re-enabled in the next
   releases.

   Features removed temporarily:
   * KerFS
   * Distributed threads
   * Checkpoint
   * Dynamic streams.

Kerrighed version 1.1.0-beta

   Features:
   * Port to Linux 2.6.11.

Kerrighed version 1.0.2
   Fixes:
   * Fix random signal delivering in some cases.
   * Fix useless network traffic when some applications are running
     (xosview for instance).
   * Fix some deadlock cases in the global lock mechanism.
   * Fix a kernel ooops in distributed thread brk function.
   * Fix large memory leak with Kerrighed threads and process migration.
   * Fix some distributed lock problems using Kerrighed threads.
   * Fix a memory leak in KerFS.

   Features:
   * Enable to see on which node a process is running in the top/ps commands.

Kerrighed version 1.0.1
   Fixes:
   * Remove deprecated dependancy on PERFCTNR.
   * Fix crash with global INET socket in some corner cases.
   * Fix major issue when HIGHMEM support is activated.

   Features:
   * Port to Linux 2.4.29.
   * Enable to use Kerrighed threads through LD_PRE_LOAD.

Kerrighed version 1.0.0
   Fixes:
   * Add permission check on commands "migrate" and "checkpoint".
   * Avoid process migration on a non existing node !
   * Fix up cluster wide PIDs problems for threads.
   * Fix crash during process migration with containers in some corner cases.

kerrighed-2.4.4/tools/krgcapset.c
/*
 *  Copyright (C) 2001-2006, INRIA, Universite de Rennes 1, EDF.
 *  Copyright (c) 2008 Jean Parpaillon <jean.parpaillon@kerlabs.com>
 *
 */

#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <stdlib.h>
#include <errno.h>
#include <getopt.h>

#include <kerrighed.h>

#define CAP_DEF(cap) [CAP_##cap] = #cap

const char * cap_text[CAP_SIZE] = {
  CAP_DEF(CHANGE_KERRIGHED_CAP),
  CAP_DEF(CAN_MIGRATE),
  CAP_DEF(DISTANT_FORK),
  CAP_DEF(FORK_DELAY),
  CAP_DEF(CHECKPOINTABLE),
  CAP_DEF(USE_REMOTE_MEMORY),
  CAP_DEF(USE_INTRA_CLUSTER_KERSTREAMS),
  CAP_DEF(USE_INTER_CLUSTER_KERSTREAMS),
  CAP_DEF(USE_WORLD_VISIBLE_KERSTREAMS),
  CAP_DEF(SEE_LOCAL_PROC_STAT),
  CAP_DEF(DEBUG),
  CAP_DEF(SYSCALL_EXIT_HOOK)
};

static struct option long_options[] = {
  {"help", no_argument, NULL, 'h'},
  {"show", no_argument, NULL, 's'},
  {"force", no_argument, NULL, 'f'},
  {"pid", required_argument, NULL, 'k'},
  {"permitted", required_argument, NULL, 'p'},
  {"effective", required_argument, NULL, 'e'},
  {"inheritable-permitted", required_argument, NULL, 'i'},
  {"inheritable-effective", required_argument, NULL, 'd'},
  {NULL, 0, NULL, 0}
};

static int supported_caps;

int print_capability(int current_cap_vector)
{
  int i ;
  int char_printed_on_this_line = 8 ;
  int char_to_print = 0 ;
  int lines_printed = 0 ;
  for (i=0; i < CAP_SIZE ;i++)
    {
      if (i == CAP_CHECKPOINTABLE)
	continue;

      if (cap_raised(current_cap_vector, i))
	{
	  char_to_print = strlen(cap_text[i])+2;
	  if (char_printed_on_this_line+char_to_print < 80)
	    {
	      if (char_printed_on_this_line!=8)
		{
		  printf(", ") ;
		  char_to_print += 2 ;
		}
	      else
		printf("\t") ;
	    }
	  else
	    {
	      printf("\n\t") ;
	      lines_printed++ ;
	      char_printed_on_this_line = 8 ;
	    }
	  printf("%s",cap_text[i]) ;
	  char_printed_on_this_line+=char_to_print ;
	}
    }
  printf("\n") ;
  return 0 ;
}


int print_capabilities(krg_cap_t * current_caps)
{
  printf("Permitted Capabilities: 0%o\n", krg_cap_getpermitted(current_caps));
  print_capability(krg_cap_getpermitted(current_caps)) ;

  printf("Effective Capabilities: 0%o\n", krg_cap_geteffective(current_caps));
  print_capability(krg_cap_geteffective(current_caps)) ;

  printf("Inheritable Permitted Capabilities: 0%o\n", krg_cap_getinheritable_permitted(current_caps));
  print_capability(krg_cap_getinheritable_permitted(current_caps)) ;

  printf("Inheritable Effective Capabilities: 0%o\n", krg_cap_getinheritable_effective(current_caps));
  print_capability(krg_cap_getinheritable_effective(current_caps)) ;

  return 0 ;
}



int construct_capability(char * description)
{
  int res = 0;
  int text_matched ;
  int i ;
  char ** next_token = &description ;
  char * current_token ;
  while (*next_token)
    {
      current_token = strsep (next_token, ",") ;
      text_matched = 0 ;
      for (i=0; i < CAP_SIZE ;i++)
	{
          if (i == CAP_CHECKPOINTABLE)
	    continue;

	  if (!strncmp(cap_text[i],current_token, strlen(cap_text[i])))
	    {
	      res |= 1<<i ;
	      text_matched = 1 ;
	      break ;
	    }
	}
      if (!text_matched)
	{
	  printf("Unrecognized capability: %s\n", current_token) ;
	  printf("Authorized capability names are:\n") ;
	  print_capability(supported_caps) ;
	}
    }
  return res ;
}



int update_capability(char* description, int oldcapability)
{
  int res, tmp ;
  
  switch (description[0])
    {
    case '+' :
      tmp = construct_capability(&description[1]) ;
      res = tmp | oldcapability ;
      break ;
    case '-' :
      tmp = construct_capability(&description[1]) ;
      res = ~tmp & oldcapability ;
      break ;
    default:
      res = construct_capability(description) ;
    }
  return res ;
}

int usage(char * argv[])
{
  printf("\
Usage: %s -h|--help\n\
  or:  %s -s|--show\n\
  or:  %s [-f|--force] [-k|--pid <pid>] {SET {[+|-]CAPABILITY LIST | OCTAL VALUE}...}\n\
", argv[0], argv[0], argv[0]);
  fputs("\n\
 -h, --help                    display this help and exit\n\
 -s, --show                    show capabilities of calling process\n\
", stdout);
  fputs("\n\
Change capabilities of the calling (or designated) process.\n\
 SET is one of:\n\
  -e, --effective              set up effective capabilities\n\
  -p, --permitted              set up permitted capabilities\n\
  -d, --inheritable-effective  set up inheritable effective capabilities\n\
  -i, --inheritable-permitted  set up inheritable permitted capabilities\n\
\n\
 CAPABILITY LIST is a comma-separated list of capabilities amongst:\n\
", stdout);
  print_capability(supported_caps);
  return 0 ;
}

int force = 0 ;
int changed_father_cap = 0 ; 
int main (int argc, char * argv[])
{
  pid_t pid = -1 ;
  int c ;
  int res ;
  long int cap_value;
  char * ignored = NULL ;
  krg_cap_t initial_caps ;

  res = krg_cap_get_supported(&supported_caps);
  if (res)
    {
      fprintf(stderr, "Unable to get the list of supported caps.\n"
	      "Is Kerrighed running?\n");
      exit(res);
    }
  if (~((1 << CAP_SIZE) - 1) & supported_caps)
    fprintf(stderr, "%s version is too old for the running kernel.\n"
		    "Some capability names won't be recognized.\n",
		    argv[0]);

  res = krg_father_capget(&initial_caps) ;
  
  if (res)
    {
      printf ("Unable to get my father's capabilities res: %d, errno: %d\n", res, errno) ;
      exit(res);
    }

  //don't manage inheritable_effective at first
  //initial_caps.krg_cap_inheritable_effective = ~0 ;

  if (argc == 1)
    {
      usage(argv);
      exit(EXIT_SUCCESS);
    }

  while (1)
    {
      int option_index = 0;
      c = getopt_long(argc, argv, "hsfk:p:e:i:d:",
		      long_options, &option_index);

      if (c == -1)
	break;

      cap_value = 0;
      switch (c) 
	{
	case 'f':
	  force = 1 ;
	  break;
	case 's':
	  print_capabilities(&initial_caps) ;
	  break;
	case 'e':
	  cap_value = strtol(optarg, &ignored, 0);
	  if (*ignored != '\0')
	    cap_value = update_capability(optarg, initial_caps.krg_cap_effective);
	  initial_caps.krg_cap_effective = cap_value;
	  changed_father_cap = 1 ;
	  break;
	case 'p':
	  cap_value = strtol(optarg, &ignored, 0);
	  if (*ignored != '\0')
	    cap_value = update_capability(optarg, initial_caps.krg_cap_permitted);
	  initial_caps.krg_cap_permitted = cap_value;
	  changed_father_cap = 1 ;
	  break;
	case 'i':
	  cap_value = strtol(optarg, &ignored, 0);
	  if (*ignored != '\0')
	    cap_value = update_capability(optarg, initial_caps.krg_cap_inheritable_permitted);
	  initial_caps.krg_cap_inheritable_permitted = cap_value;
	  changed_father_cap = 1 ;
	  break;
	case 'd':
	  cap_value = strtol(optarg, &ignored, 0);
	  if (*ignored != '\0')
	    cap_value = update_capability(optarg, initial_caps.krg_cap_inheritable_effective);
	  initial_caps.krg_cap_inheritable_effective = cap_value;
	  changed_father_cap = 1 ;
	  break;
	case 'k':
	  pid = atoi (optarg) ;
	  if (changed_father_cap)
	    {
	      printf("Conflict in implementation: -k must be the first option used\n") ;
	      exit(EXIT_FAILURE) ;
	    }
	  res = krg_pid_capget(pid, &initial_caps) ;
	  if (res)
	    {
	      printf ("Unable to get capabilities of process %d (res: %d, errno: %d))\n", pid, res, errno) ;
	      exit(res) ;
	    }
	  break;
	case 'h':
	  usage(argv);
	  exit(EXIT_SUCCESS);
	  break;
	default:
	  usage(argv);
	  exit(EXIT_FAILURE);
	  break;
	}
    }
  if (changed_father_cap && (!force && (!cap_raised(initial_caps.krg_cap_permitted, CAP_CHANGE_KERRIGHED_CAP) ||
					!cap_raised(initial_caps.krg_cap_effective, CAP_CHANGE_KERRIGHED_CAP) )))
    {
      int answer ;
      int use_default = 1;
      printf("Are you sure you want to loose the capability to change your capabilities (y/N)\n") ;
      answer = getchar() ;
      while (answer != 'y' && answer != 'Y' && answer != 'n' && answer != 'N' && answer != EOF && answer != '\n')
	{
	  use_default = 0;
	  answer = getchar();
	}
      if (answer != 'y' && answer != 'Y' && (answer = '\n' || use_default) )
	exit(EXIT_SUCCESS);
    }
  if (changed_father_cap)
    {
      if (pid == -1)
	{
	  res = krg_father_capset(&initial_caps) ;
	  if (res)
	    printf ("Unable to set my father's capabilities\n") ;
	}
      else
	{
	  res = krg_pid_capset(pid, &initial_caps) ;
	  if (res)
	    printf ("Unable to set the capabilities of process %d (res is %d, errno %d)\n", pid, res, errno) ;
	}
    }
  exit(EXIT_SUCCESS);
}
